webapp/src/components/SingleStudy/MapView/index.tsx

```diff
@@ -0,0 +1,391 @@
+      source,
+      target,
+    };
+    setSelectedItem(obj);
+  }, []);
+
+  const createModeLink = useCallback((id: string) => {
+    if (firstNode && firstNode === id) {
+      setFirstNode(undefined);
+      setSecondNode(undefined);
+    } else {
+      setFirstNode(id);
+    }
+  }, [firstNode, setFirstNode, setSecondNode]);
+
+  const onClose = () => {
+    setOpenModal(false);
+  };
+
+  const onSave = async (name: string, posX: number, posY: number, color: string) => {
+    setOpenModal(false);
+    try {
+      await createArea(study.id, name);
+      setNodeData([...nodeData, {
+        id: name,
+        x: posX,
+        y: posY,
+        color,
+        rgbColor: color.slice(4, -1).split(',').map(Number),
+        size: { width: calculateSize(name), height: NODE_HEIGHT },
+      }]);
+    } catch (e) {
+      setNodeData(nodeData.filter((o) => o.id !== name));
+      enqueueErrorSnackbar(enqueueSnackbar, t('singlestudy:createAreaError'), e as AxiosError);
+    }
+  };
+
+  const handleUpdate = async (id: string, value: UpdateAreaUi) => {
+    const filterNodes = nodeData.find((o) => o.id === id);
+    if (filterNodes) {
+      try {
+        const prevColors = filterNodes.rgbColor;
+        if (value.color_rgb[0] !== prevColors[0] || value.color_rgb[1] !== prevColors[1] || value.color_rgb[2] !== prevColors[2]) {
+          const updateNode = nodeData.filter((o) => o.id !== id);
+          setNodeData([...updateNode, {
+            id,
+            x: value.x,
+            y: value.y,
+            color: `rgb(${value.color_rgb[0]}, ${value.color_rgb[1]}, ${value.color_rgb[2]})`,
+            rgbColor: filterNodes.rgbColor,
+            size: filterNodes.size,
+          }]);
+          await updateAreaUI(study.id, id, value);
+        }
+      } catch (e) {
+        setNodeData([...nodeData]);
+        enqueueErrorSnackbar(enqueueSnackbar, t('singlestudy:colorUpdateError'), e as AxiosError);
+      }
+    }
+  };
+
+  const handleUpdatePosition = async (id: string, x: number, y: number) => {
+    const filterNodes = nodeData.find((o) => o.id === id);
+    if (filterNodes) {
+      try {
+        const prevPosition = { x: nodeData.find((o) => o.id === id)?.x, y: nodeData.find((o) => o.id === id)?.y };
+        if (x !== prevPosition.x || y !== prevPosition.y) {
+          const updateNode = nodeData.filter((o) => o.id !== id);
+          setNodeData([...updateNode, {
+            id,
+            x: Math.floor(x),
+            y: Math.floor(y),
+            color: filterNodes.color,
+            rgbColor: filterNodes.rgbColor,
+            size: filterNodes.size,
+          }]);
+          // eslint-disable-next-line @typescript-eslint/camelcase
+          await updateAreaUI(study.id, id, { x, y, color_rgb: filterNodes.rgbColor });
+        }
+      } catch (e) {
+        setNodeData([...nodeData]);
+        enqueueErrorSnackbar(enqueueSnackbar, t('singlestudy:positionUpdateError'), e as AxiosError);
+      }
+    }
+  };
```

Comment:
you can factor this code with the previous function, just call it with a UpdateUI object that you will create from the x,y and targetNode older properties